diff --git a/mlx/c/ops.cpp b/mlx/c/ops.cpp
index bf565d1..7c4f57b 100644
--- a/mlx/c/ops.cpp
+++ b/mlx/c/ops.cpp
@@ -1708,6 +1708,7 @@ extern "C" int mlx_gather_qmm(
             (bits.has_value ? std::make_optional<int>(bits.value)
                             : std::nullopt),
             std::string(mode),
+            std::nullopt,
             sorted_indices,
             mlx_stream_get_(s)));
   } catch (std::exception& e) {
@@ -1716,6 +1716,50 @@ extern "C" int mlx_gather_qmm(
   }
   return 0;
 }
+extern "C" int mlx_gather_qmm_with_global_scale(
+    mlx_array* res,
+    const mlx_array x,
+    const mlx_array w,
+    const mlx_array scales,
+    const mlx_array biases /* may be null */,
+    const mlx_array lhs_indices /* may be null */,
+    const mlx_array rhs_indices /* may be null */,
+    bool transpose,
+    mlx_optional_int group_size,
+    mlx_optional_int bits,
+    const char* mode,
+    const mlx_array global_scale,
+    bool sorted_indices,
+    const mlx_stream s) {
+  try {
+    mlx_array_set_(
+        *res,
+        mlx::core::gather_qmm(
+            mlx_array_get_(x),
+            mlx_array_get_(w),
+            mlx_array_get_(scales),
+            (biases.ctx ? std::make_optional(mlx_array_get_(biases))
+                        : std::nullopt),
+            (lhs_indices.ctx ? std::make_optional(mlx_array_get_(lhs_indices))
+                             : std::nullopt),
+            (rhs_indices.ctx ? std::make_optional(mlx_array_get_(rhs_indices))
+                             : std::nullopt),
+            transpose,
+            (group_size.has_value ? std::make_optional<int>(group_size.value)
+                                  : std::nullopt),
+            (bits.has_value ? std::make_optional<int>(bits.value)
+                            : std::nullopt),
+            std::string(mode),
+            (global_scale.ctx ? std::make_optional(mlx_array_get_(global_scale))
+                              : std::nullopt),
+            sorted_indices,
+            mlx_stream_get_(s)));
+  } catch (std::exception& e) {
+    mlx_error(e.what());
+    return 1;
+  }
+  return 0;
+}
 extern "C" int mlx_gather_qqmm(
     mlx_array* res,
     const mlx_array x,
diff --git a/mlx/c/ops.h b/mlx/c/ops.h
index 7417785..4323212 100644
--- a/mlx/c/ops.h
+++ b/mlx/c/ops.h
@@ -550,6 +550,21 @@ int mlx_gather_qmm(
     const char* mode,
     bool sorted_indices,
     const mlx_stream s);
+int mlx_gather_qmm_with_global_scale(
+    mlx_array* res,
+    const mlx_array x,
+    const mlx_array w,
+    const mlx_array scales,
+    const mlx_array biases /* may be null */,
+    const mlx_array lhs_indices /* may be null */,
+    const mlx_array rhs_indices /* may be null */,
+    bool transpose,
+    mlx_optional_int group_size,
+    mlx_optional_int bits,
+    const char* mode,
+    const mlx_array global_scale,
+    bool sorted_indices,
+    const mlx_stream s);
 int mlx_gather_qqmm(
     mlx_array* res,
     const mlx_array x,
