From 06e9f95ae90a7de7463718a85206d1c85da11213 Mon Sep 17 00:00:00 2001
From: Liu Yuan <namei.unix@gmail.com>
Date: Sun, 13 Oct 2024 01:35:50 +0800
Subject: [PATCH] 01-load-progress.diff

---
 common/common.cpp | 2 ++
 common/common.h   | 6 ++++++
 2 files changed, 8 insertions(+)

diff --git a/common/common.cpp b/common/common.cpp
index 451307b5..6bc11614 100644
--- a/common/common.cpp
+++ b/common/common.cpp
@@ -999,6 +999,8 @@ struct llama_model_params common_model_params_to_llama(const common_params & par
     mparams.use_mmap        = params.use_mmap;
     mparams.use_mlock       = params.use_mlock;
     mparams.check_tensors   = params.check_tensors;
+    mparams.progress_callback = params.progress_callback;
+    mparams.progress_callback_user_data = params.progress_callback_user_data;
     if (params.kv_overrides.empty()) {
         mparams.kv_overrides = NULL;
     } else {
diff --git a/common/common.h b/common/common.h
index 71e68615..e54b1f6f 100644
--- a/common/common.h
+++ b/common/common.h
@@ -265,6 +265,12 @@ struct common_params {
     // multimodal models (see examples/llava)
     std::string mmproj = "";        // path to multimodal projector                                         // NOLINT
     std::vector<std::string> image; // path to image file(s)
+    // Called with a progress value between 0.0 and 1.0. Pass NULL to disable.
+    // If the provided progress_callback returns true, model loading continues.
+    // If it returns false, model loading is immediately aborted.
+    llama_progress_callback progress_callback = NULL;
+    // context pointer passed to the progress callback
+    void * progress_callback_user_data;
 
     // embedding
     bool embedding         = false; // get only sentence embedding
-- 
2.34.1

