diff --git a/examples/server/server.cpp b/examples/server/server.cpp
index 895d608f..07ca2463 100644
--- a/examples/server/server.cpp
+++ b/examples/server/server.cpp
@@ -1153,12 +1153,13 @@ struct server_context {
                 // add the token to slot queue and cache
             }
 
-            slot.add_token_string(result);
             if (slot.params.stream) {
                 send_partial_response(slot, result);
             }
         }
 
+        slot.add_token_string(result);
+
         if (incomplete) {
             slot.has_next_token = true;
         }
