From 6465fec6290f0a7f5d4d0fbe6bcf634e4810dde6 Mon Sep 17 00:00:00 2001
From: Michael Yang <mxyng@pm.me>
Date: Mon, 23 Oct 2023 10:39:34 -0700
Subject: [PATCH] default log stderr

---
 common/log.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/common/log.h b/common/log.h
index b8953fd..25522cd 100644
--- a/common/log.h
+++ b/common/log.h
@@ -90,7 +90,7 @@
 //  }
 //
 #ifndef LOG_TARGET
-    #define LOG_TARGET log_handler()
+    #define LOG_TARGET nullptr
 #endif
 
 #ifndef LOG_TEE_TARGET
-- 
2.42.0

