diff --git a/tools/server/CMakeLists.txt b/tools/server/CMakeLists.txt
index 02607c838..8d8a6e814 100644
--- a/tools/server/CMakeLists.txt
+++ b/tools/server/CMakeLists.txt
@@ -2,7 +2,7 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
 
 # MSVC emits a PCH bookkeeping symbol that WINDOWS_EXPORT_ALL_SYMBOLS exports as an ambiguous "__"
 
-set(LLAMA_SERVER_PCH ON)
+set(LLAMA_SERVER_PCH OFF)
 
 if (BUILD_SHARED_LIBS AND CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
     set(LLAMA_SERVER_PCH OFF)
