diff --git a/mlx/c/ops.cpp b/mlx/c/ops.cpp
index bf565d1..435706a 100644
--- a/mlx/c/ops.cpp
+++ b/mlx/c/ops.cpp
@@ -1687,6 +1687,7 @@ extern "C" int mlx_gather_qmm(
     mlx_optional_int group_size,
     mlx_optional_int bits,
     const char* mode,
+    const mlx_array global_scale /* may be null */,
     bool sorted_indices,
     const mlx_stream s) {
   try {
@@ -1708,6 +1709,8 @@ extern "C" int mlx_gather_qmm(
             (bits.has_value ? std::make_optional<int>(bits.value)
                             : std::nullopt),
             std::string(mode),
+            (global_scale.ctx ? std::make_optional(mlx_array_get_(global_scale))
+                              : std::nullopt),
             sorted_indices,
             mlx_stream_get_(s)));
   } catch (std::exception& e) {
diff --git a/mlx/c/ops.h b/mlx/c/ops.h
index 7417785..6dfbf9e 100644
--- a/mlx/c/ops.h
+++ b/mlx/c/ops.h
@@ -548,6 +548,7 @@ int mlx_gather_qmm(
     mlx_optional_int group_size,
     mlx_optional_int bits,
     const char* mode,
+    const mlx_array global_scale /* may be null */,
     bool sorted_indices,
     const mlx_stream s);
 int mlx_gather_qqmm(
